wagtail/wagtaildocs/views/serve.py

```diff
@@ -13,8 +17,40 @@ def serve(request, document_id, document_filename):
     # Send document_served signal
     document_served.send(sender=Document, instance=doc, request=request)
 
-    if hasattr(settings, 'SENDFILE_BACKEND'):
-        return sendfile(request, doc.file.path, attachment=True, attachment_filename=doc.filename)
+    try:
+        local_path = doc.file.path
+    except NotImplementedError:
+        local_path = None
+
+    if local_path:
+
+        # Use wagtail.utils.sendfile to serve the file;
+        # this provides support for mimetypes, if-modified-since and django-sendfile backends
+
+        if hasattr(settings, 'SENDFILE_BACKEND'):
+            return sendfile(request, local_path, attachment=True, attachment_filename=doc.filename)
+        else:
+            # Fallback to streaming backend if user hasn't specified SENDFILE_BACKEND
+            return sendfile(request, local_path, attachment=True, attachment_filename=doc.filename, backend=sendfile_streaming_backend.sendfile)
+
     else:
-        # Fallback to streaming backend if user hasn't specified SENDFILE_BACKEND
-        return sendfile(request, doc.file.path, attachment=True, attachment_filename=doc.filename, backend=sendfile_streaming_backend.sendfile)
+
+        # We are using a storage backend which does not expose filesystem paths
+        # (e.g. storages.backends.s3boto.S3BotoStorage).
+        # Fall back on pre-sendfile behaviour of reading the file content and serving it
+        # as a StreamingHttpResponse
+
+        wrapper = FileWrapper(doc.file)
+        response = StreamingHttpResponse(wrapper, content_type='application/octet-stream')
+
+        try:
+            response['Content-Disposition'] = 'attachment; filename=%s' % doc.filename
+        except BadHeaderError:
+            # Unicode filenames can fail on Django <1.8, Python 2 due to
+            # https://code.djangoproject.com/ticket/20889 - try with an ASCIIfied version of the name
+            response['Content-Disposition'] = 'attachment; filename=%s' % unidecode(doc.filename)
+
+        # FIXME: storage backends are not guaranteed to implement 'size'
```

Comment:
Potentially, yes - however, I'd have to check whether leaving out the `Content-Length` header is valid for a streaming response (and I have a sneaking feeling it isn't). Dealing with that would be out of scope for this bugfix - I've stuck to the same code that existed in 1.0b1, so in the event of someone using a storage backend without the `size` property (FWIW, the Boto S3 backend *does* implement it), it will be no more broken than it was in earlier versions :-)